module/x/gravity/keeper/evidence.go

```diff
@@ -0,0 +1,93 @@
+package keeper
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/cosmos/gravity-bridge/module/x/gravity/types"
+)
+
+func (k Keeper) CheckBadSignatureEvidence(
+	ctx sdk.Context,
+	msg *types.MsgSubmitBadSignatureEvidence) error {
+	var subject types.EthereumSigned
+
+	k.cdc.UnpackAny(msg.Subject, &subject)
+
+	switch subject := subject.(type) {
+	case *types.OutgoingTxBatch:
+		return k.checkBadSignatureEvidenceInternal(ctx, subject, msg.Signature)
+	case *types.Valset:
+		return k.checkBadSignatureEvidenceInternal(ctx, subject, msg.Signature)
+	case *types.OutgoingLogicCall:
+		return k.checkBadSignatureEvidenceInternal(ctx, subject, msg.Signature)
```

Comment:
I don't understand why you do you have the switch in the first place if you are calling the same function?

Reply:
The logic needs to be run, but only if the subject is one of those 3 types. This pattern was suggested by @jackzampolin on Tuesday

Reply:
I think @jackzampolin suggested to use an interface and remove the switch, which is what makes sense in this case